pkg/metrics/node.go

```diff
@@ -0,0 +1,179 @@
+	mgmtcontrollers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/wrangler/pkg/ticker"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	nodeProviderLabel  = "provider"
+	nodeClusterIDLabel = "cluster_id"
+	nodeIsActiveLabel  = "is_active"
+)
+
+var (
+	nodeLabels = []string{nodeProviderLabel, nodeClusterIDLabel, nodeIsActiveLabel}
+	numNodes   = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "nodes",
+			Help:      "Number of nodes in rancher managed clusters",
+		}, nodeLabels,
+	)
+	numCores = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "node_cores",
+			Help:      "Number of node cores in rancher managed clusters",
+		}, nodeLabels,
+	)
+)
+
+const (
+	reportInterval = time.Duration(1 * time.Minute)
+	logPrefix      = "[prometheus-node-metrics]"
+)
+
+type nodeMetrics struct {
+	nodeCache    mgmtcontrollers.NodeCache
+	clusterCache mgmtcontrollers.ClusterCache
+}
+
+func (m *nodeMetrics) collect(ctx context.Context) {
+	for range ticker.Context(ctx, reportInterval) {
+		logrus.Debugf("%s collecting nodes to report metrics", logPrefix)
+
+		// list all nodes
+		nodes, err := m.nodeCache.List("", labels.NewSelector())
+		if err != nil {
+			logrus.Errorf("%s couldn't list v3.Nodes: %v", logPrefix, err)
+			continue
+		}
+
+		var infos []*nodeInfo
+		for _, node := range nodes {
+			info, err := m.getNodeInfo(node)
+			if err != nil {
+				logrus.Debugf("%s could not determine node info: %v", logPrefix, err)
+				continue
+			}
+			infos = append(infos, info)
+		}
+
+		setMetrics(infos)
+	}
+
+	logrus.Debugf("%s context cancelled, exiting", logPrefix)
+}
+
+type nodeLabelValues struct {
+	provider  string
+	clusterID string
+	isActive  bool
+}
+
+type nodeInfo struct {
+	coreCount   int
+	labelValues nodeLabelValues
+}
+
+// getNodeInfo uses the v3.Node object and it's corresponding v3.Cluster object to get metric info for a node
+func (m *nodeMetrics) getNodeInfo(node *v3.Node) (*nodeInfo, error) {
+	cluster, err := m.clusterCache.Get(node.ObjClusterName())
+	if err != nil {
+		return nil, fmt.Errorf("error getting cluster %s for node %s, %v", node.ObjClusterName(), node.Name, err)
+	}
+	if cluster == nil {
+		return nil, fmt.Errorf("could not get associated cluster for node: %s/%s", node.Namespace, node.Name)
+	}
+	if cluster.DeletionTimestamp != nil {
+		return nil, fmt.Errorf("associated cluster for node: %s/%s is being deleted", node.Namespace, node.Name)
+	}
+
+	var isActive bool
+	// node conditions aren't reliably updated when node is unreachable so use the cluster ready condition
+	// if the cluster is ready the its nodes are considered to be active
+	if v32.ClusterConditionReady.IsTrue(cluster) {
+		isActive = true
```

Comment:
I know that we wanted to avoid making this logic overly complex. That being said, do we want to do some sort of sanity check on the node so that both the node and the cluster are ready? I'm thinking something like:
```
isActive = v32.ClusterConditionReady.IsTrue(cluster) && node.Status.InternalNodeStatus.Phase == types.NodeRunning
```
You could also loop through the node conditions and make sure that there is a condition of type `NodeReady`, but I think that's a bit more complex than you were looking for (looks like conditions are current, so that wouldn't be too bad, but still).

Reply:
I initially had that, but the node doesn't set it's ready condition to false when expected.

When I checked both I never saw a false value come through when bringing down nodes removing agents, etc. So in effect this was like not having the label at all (since the value was always `true`)

Worth mentioning that all nodes in a given cluster will either all be active or none with this change though. I think that's acceptable and I'm also getting provider from the cluster object

Reply:
One additional note on this. When provisioning a cluster, node metrics for nodes in that cluster won't show until the cluster is provisioned. The reason for this is that the `provider` comes from the `cluster.Status.Provider` field, which doesn't get populated until provisioned. If there is no provider, we don't update node metrics. 

Another possible option to include these nodes would be to add an "unknown" provider value count nodes from clusters that don't have a provider in their status (yet)

Reply:
I made the change to include an `unknown` provider label value when the cluster doesn't have the provider value yet. This allows provisioning nodes to be captured in the metrics

Reply:
I've updated this to check the node too